### ISSUE TITLE
Garbage collection for tqdm._instances

### ISSUE THREAD

**Issue by skjerns** [tqdm/tqdm#1025]
As pointed out in the FAQ, there seem to be multiple issues with newlines appearing in custom consoles (e.g. qtconsole/Spyder), or maybe even any kind of interactive console.

I've figured out, that they do not appear when using `position=0` and the problem seems to be that any code that is run and interrupted with a `KeyboardInterrupt` will leave a dead loop references in `tqdm._instances`.

Would it be to implement some sort of garbage collection mechanism that can detect these unreachable instances? E.g. counting the refcount from `gc`?

**Comment by casperdcl** [comment]
Does this still occur with [tqdm>=4.45.0](https://tqdm.github.io/releases/#v4450-2020-04-02)?

To be clear, there are 3 cases and the last one is cumbersome, un-Pythonic and probably something which should not be automatically handled without requiring `finally`:

```python
for i in tqdm(iterable):
	# case 1: should be fine to interrupt

with tqdm(total=len(iterable)) as t:
	for i in iterable:
		# case 2: should be fine to interrupt

t = tqdm(total=len(iterable))
try:
	for i in iterable:
		# case 3: should be fine to interrupt
finally:
	t.close()
```

**Comment by skjerns** [comment]
Exactly, the third case is what I have. You're right, I can solve this with a try/finally statement

```Python
def compute():
    loop = tqdm(total=10)
    for i in range(10):
        loop.update()
        raise KeyboardInterrupt


for i in tqdm(range(10)):
    compute()
print(len(tqdm._instances)) # 1
```

However, in this case the loop object would be unreachable from outside the function, and would actually be collected by the `gc`, if not for the reference in `tqdm._instances`. This is quite a minor bug, however it might be solvable?

great work by the way, tqdm is awesome.

**Comment by casperdcl** [comment]
> ```python
> def compute():
>     loop = tqdm(total=10)
>     for i in range(10):
>         loop.update()
>         raise KeyboardInterrupt
> ```

rather than adding `finally`, it's better to do:

```python
def compute():
    with tqdm(total=10) as loop:
        for i in range(10):
            loop.update()
            raise KeyboardInterrupt
```

> However, in this case the loop object would be unreachable from outside the function

you could return the `loop` object from the function if that's what you want.

> would actually be collected by the `gc`, if not for the reference in `tqdm._instances`. This is quite a minor bug, however it might be solvable? Anyway, I don't even understand how you figure out that the loop is finished when there is no interrupt, so I don't have a clue if that is easily possible.

no. It would be collected if you:

- exhaust a `tqdm`-wrapped iterable, or
- catch an exception (automatic using the `with` syntax or manually using `finally: close()`), or
- `del` the `tqdm` object

> question on the side: is there a convenience-function to get a handle to the latest tqdm loop? (besides a bit cumbersome `list(tqdm._instances)[-1]`), this would make it easier to e.g. set a prefix without passing the handle to a subfunction

no; if you need to keep track of multiple `tqdm` objects, the current recommended way is to manage your own list of them. If you want to use the builtin private set, it's probably better with use `max(tqdm._instances)`
